### PR TITLE
BUG: groupby.nth should be a filter

### LINKED ISSUE #13666: groupby.nth() labelling conventions changed from 0.17 -> 0.18
In pandas 0.17.1, grouping a DatetimeIndex with monthly frequency produced labels keyed at the start of the month, and `nth()` produces a label keyed by the selected row.  In 0.18, we unconditionally return groups labelled by the end of the month.
#### Old Pandas

```
In [25]: pd.__version__
Out[25]: u'0.17.1'

In [26]: dates = pd.date_range('2014', periods=90)

In [27]: pd.Series(index=dates).groupby(pd.TimeGrouper(freq='M')).nth(5)
Out[27]:
2014-01-06   NaN
2014-02-06   NaN
2014-03-06   NaN
dtype: float64
```
#### Latest Pandas

```
In [71]: pd.__version__
Out[71]: u'0.18.1'

In [72]: dates = pd.date_range('2014', periods=90)

In [73]: pd.Series(index=dates).groupby(pd.TimeGrouper(freq='M')).nth(5)
Out[73]:
2014-01-31   NaN
2014-02-28   NaN
2014-03-31   NaN
dtype: float64
```
#### Expected Output

I expect the behavior from 0.17.1.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #13666 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -3001,97 +3001,67 @@ def nth(
...
         >>> g.nth(0, dropna='any')
-             B
-        A
-        1  2.0
-        2  3.0
+           A   B
+        1  1 2.0
+        2  2 3.0
 
         NaNs denote group exhausted when using dropna
```
Comment (mroeschke) [inline pandas/core/groupby/groupby.py:3048]: This description probably needs updating

File: doc/source/whatsnew/v2.0.0.rst [hunk 2]
```diff
@@ -338,7 +338,7 @@ Groupby/resample/rolling
 - Bug in :meth:`DataFrameGroupBy.sample` raises ``ValueError`` when the object is empty (:issue:`48459`)
 - Bug in :meth:`Series.groupby` raises ``ValueError`` when an entry of the index is equal to the name of the index (:issue:`48567`)
 - Bug in :meth:`DataFrameGroupBy.resample` produces inconsistent results when passing empty DataFrame (:issue:`47705`)
--
+- Bug in :meth:`.DataFrameGroupBy.nth` and :meth:`.SeriesGroupBy.nth` would treat operation as a aggregation whereas it is a filtration; in particular, the result index no longer contains the groupers but rather is filtered from the original index of the input (:issue:`13666`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.0.rst:341]: Generally like consistency now that `nth` is a filter, but I think this should be called out in it's own "notable bug fix" section

**Comment by rhshadrach** [comment]
Thanks @mroeschke - ready for another look.

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -3001,97 +3001,68 @@ def nth(
...
-        2  3.0
+           A   B
+        1  1 2.0
+        2  2 3.0
 
-        NaNs denote group exhausted when using dropna
+        When the specified ``n`` is larger than any of the groups, an
+        empty DataFrame is returned
 
         >>> g.nth(3, dropna='any')
```
Comment (mroeschke) [inline pandas/core/groupby/groupby.py:3051]: Do you think this example should be shown in the whatsnew? On the surface, this example appears to be quite different from before

**Comment by WillAyd** [comment]
Hmm I'm not sure I agree with this approach - doesn't this introduce an even larger inconsistency now between nth / first / last? Do we have other groupby functions that work this way?

**Comment by rhshadrach** [comment]
> Hmm I'm not sure I agree with this approach - doesn't this introduce an even larger inconsistency now between nth / first / last? Do we have other groupby functions that work this way?

`first` and `last` are reductions - they return 1 value per group. `nth` on the other hand can also return zero or multiple rows. It is because of this I think we should be treating `nth` as a filtration. The other filtrations, `head`, `tail`, and `filter` all behave as in this PR, namely they merely filter the rows from the input object. In particular, `groupby(...).head()` now behaves the same as `groupby(...).nth([0, 1, 2, 3, 4])`.

**Comment by WillAyd** [comment]
Yea I definitely see the argument. I think I'm +/-0 . IIUC this makes nth closer to filtering on rank, but with a predetermined option for a tiebreaker

**Comment by rhshadrach** [comment]
> IIUC this makes nth closer to filtering on rank

nth preserves the input order, making it somewhat different from rank.

**Review by mroeschke — APPROVED** [review]
I'm okay with this change as nth as a filter makes sense to me and to align the corresponding output. IIRC there was an issue to write `groupby().head/tail` in terms with `nth` which this change would help IIUC.

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Awesome work @rhshadrach

**Pull request merged**